### ISSUE TITLE
Handling multiple roles on the graphql client (possibly in the same query)

### ISSUE THREAD

**Issue by revskill10** [hasura/graphql-engine#877]
- In an application, one user often have multi roles at the same time, at a screen.
Let's say, in a blog application, both editor and author could edit a post. So if a user is both an editor and an author, how to authenticate with hasura then ? 
In this case, hasura should support multi-roles in headers: `x-hasura-roles` instead of `x-hasura-role`.
Else how do you decide which role the user should use ? Because all roles are equal and valid.

- What i mean by multi-roles query is that.
```
query {
table1 {
 field1
}
table2 {
 field2
}
```
Suppose that user has roles role1 for table1, and role2 for table2 .
This is currently impossible to make above query.

- OK, what's if a websocket connection is made by above query. Now the user's role changed, what happen ?
The websocket connection should not be recreated for that. Instead the subscription data should reflect the change after that.

And this is the second hard part to solve this problem (when roles change for a subscription query)

- Suggested implementation (per @coco98 )
The query now should look like this
```
query {
  asUser {
    someTable {...}
  }
  asManager {
    someTable {...}
  }
}
```

## Note 2 : My proposed solution
In Console UI for permission setting, there must be ability to assign current permission to another roles in system (or creation new role).
Only then that, Hasura could validate one role or array of roles query to resolve correct data to return.

**Comment by Maxpain177** [comment]
+1. Same problem

**Comment by revskill10** [comment]
One more note: This feature allows you to put multi-roles queries in a HTTP graphql request.
For subscription to work, hasura needs one more thing:
- It requires a pool of websocket connections which could handle any roles that are sent to it.

**Comment by go4cas** [comment]
@shahidhk ... my understanding of the current JWT role workflow is:

1. The token is decoded, and checked whether both `x-hasura-default-role` and `x-hasura-allowed-roles` are included
2. If both headers exist, it is checked whether the `x-hasura-default-role` is part of the roles defined in `x-hasura-allowed-roles`
3. If the `x-hasura-role` header is present, then this value is checked against the `x-hasura-allowed-roles` and it overrides `x-hasura-default-role` as the role to use
4. The role (either `x-hasura-default-role` or `x-hasura-role`) is then checked against the role defined on the permissions on the table(s) in the graphql query.

My suggestion (and it's really a question whether this is doable or not):
The `x-hasura-default-role` becomes optional, and the following is implemented:
- If both `x-hasura-default-role` **AND** `x-hasura-role` are present, use the workflow as described above
- If either `x-hasura-default-role` **OR** `x-hasura-role` is present, use whichever is available, and use the workflow as described above
- If NEITHER `x-hasura-default-role` **AND** `x-hasura-role` are present, then simply use the `x-hasura-allowed-roles` to check against the table(s) in the graphql query.

What do you think?

This way, the role system will be backwards compatible, and you give developers an option on how to implement roles coming from their client apps.

**Comment by coco98** [comment]
@go4cas What do would you expect to happen when a graphql query is valid for 2 roles? Which role should it run as, or do you expect the results to be automatically merged?

On another note, what do you think about using roles as root-fields?

```
query {
  asUser {
    someTable {...}
  }
  asManager {
    someTable {...}
  }
}
```

This way the client has easy control over knowing which data to fetch. This would work quite easily with point 3 in your suggestion where only `x-hasura-allowed-roles` and no other headers are present.

Thoughts?

**Comment by revskill10** [comment]
@go4cas The tricky (and the hardest) part of this problem is currently Hasura only allow one role graphql query. It couldn't handle multi roles query.

**Comment by revskill10** [comment]
@coco98 As i understand, in your query, hasura should automatically handle each query for each role, then resolve all to return data to client.

**Comment by coco98** [comment]
@go4cas Correct! What happens if John had the roles `['product-user', 'product-admin', 'order-user'...]`. 

Basically, what I meant to ask was, what if there are multiple applicable roles? Do both get applied, or does only the first one, or do all get applied and the results get merged?

**Comment by go4cas** [comment]
The way I think about it is that you have two sets of roles: the roles defined on the table, and the roles assign to the user. The check should be as simple as: "do any of the roles defined on the table appear in the list of roles for the user? If so, grant permission. If not, 'permission denied!'"

**Comment by go4cas** [comment]
Ah wait ... I see what you mean now. You are saying what if the table has multiple roles, each with different permissions, and John belongs to both those roles. Which permissions apply?

**Comment by Maxpain177** [comment]
In my case I should create two roles for most tables: `anonymous` and `user`, which have the same subset of fields as `anonymous` + some additional fields.

**Comment by revskill10** [comment]
@go4cas What i mean by `multi-roles query` is that.
```
query {
table1 {
 field1
}
table2 {
 field2
}
```
Suppose that user has roles `role1` for `table1`, and `role2` for `table2` .
This is currently impossible to make above query.

**Comment by Maxpain177** [comment]
I suggest adding an ability to inherit permissions from another role.

**Comment by revskill10** [comment]
OK, what's if a websocket connection is made by above query. Now the user's role changed, what happen ?
The websocket connection should not be recreated for that. Instead the subscription data should reflect the change after that.
And this is the second hard part to solve this problem (when roles change for a subscription query)

**Comment by revskill10** [comment]
> @go4cas What do would you expect to happen when a graphql query is valid for 2 roles? Which role should it run as, or do you expect the results to be automatically merged?
> 
> On another note, what do you think about using roles as root-fields?
> 
> ```
> query {
>   asUser {
>     someTable {...}
>   }
>   asManager {
>     someTable {...}
>   }
> }
> ```
> This way the client has easy control over knowing which data to fetch. This would work quite easily with point 3 in your suggestion where only `x-hasura-allowed-roles` and no other headers are present.
> 
> Thoughts?

What if the query is like that:
```
query {
asUserAndAuthor {
 table {...}
}
```
?

**Comment by coco98** [comment]
@revskill10 True, regarding websockets. However, we can discuss the how to implement and the caveats with implementation later. I was thinking of first just collecting ideas around what kind of roles systems would work best :)

Once we have more thoughts posted here, I'll post a summary of all the ideas too!

**Comment by ecthiender** [comment]
Adding few notes just to clarify few cases/scenarios:

 @coco98 already mentioned this:
> what happens when a graphql query is valid for 2 roles? Which role should it run as, or do you expect the results to be automatically merged? 

The automatic merging bit also can't be the **only** behaviour, because of some use-cases. Take the below example (albeit a bit contrived):

Let's say there's a table called `ride_details` which has the following fields:
```
ride_details:
  - passenger_name
  - passenger_personal_details
  - driver_name
  - driver_personal_details
```
Now there are two roles `passenger` and `driver` with permissions on this table. `passenger` role has access to columns `passenger_name`, `passenger_personal_details`, `driver_name` but **not** `driver_personal_details`, similarly `driver` role has permissions on `drive_name`, `driver_personal_details`, `passenger_name` but not `passenger_personal_details`.

Now if a user has both the roles `passenger` and `driver`, and if the behaviour is to send multiple roles in a query and it applies permissions of all the roles and merges the data and returns, it won't work for the above scenario. In the above scenario, you specifically would want to query as separate roles and not want merged data of both the roles.

So even if we support multiple roles in one query and the default behaviour is to merge the resultant data of two roles, there has to be (optional) mechanism to still specify an individual role.

**PS**: This was probably the main reason why we stuck to the current system and not had multiple roles in one query

**Comment by revskill10** [comment]
@ecthiender How about this query ?
```
query {
passenger_details:ride_details(role: "passenger") { passenger_personal_details }
driver_details:ride_details(role:"driver") { driver_personal_details }
```
The response will look like:
```json
{ data: { passenger_details: [...] , driver_details: [...] } }
```

**Comment by ecthiender** [comment]
@revskill10 in the actual application: when I'm taking a ride (`passenger` role), I'm not allowed to look at driver details. But I have both the roles under my user account in the application. So the client application needs to make the query only as role passenger and not make the query with driver role at all.

**Comment by revskill10** [comment]
@ecthiender For example, you're an editor, and an author.
You're in admin dashboard.
Because you're an editor, you can see `more` data than a normal author.
And because you're an author, you can fetch your own posts.

So in the posts page in `/admin` routes, you should be able to fetch multiple tables at the same time. In this case, you fetch `posts` table as an `author`, and fetch other tables as an `editor`.
Does this use case make sense ?

**Comment by ecthiender** [comment]
Yeah. This is the most common use-case for multiple roles.

But I was giving an example, where having multiple roles and **only**
automatically merging them won't work. There has to be an other option to
specify a particular role - even if we make multiple roles the default
option.

**Comment by revskill10** [comment]
@ecthiender  I have a suggestion for the problem:
When user add a permision for a table, the graphql API should generate with prefix (role name).
For example, the `posts` table should have 3 api:
```
posts {...} (as a guest)
author_posts {...} (as an author)
editor_posts {...} (as an editor)
```
How do you think ?

**Comment by revskill10** [comment]
I have a simple solution for this problem.
In Console UI for permission setting, there must be ability to assign current permission to another roles in system (or creation).
Only then that, Hasura could validate one role or array of roles query to resolve correct data to return.

**Comment by dsandip** [comment]
@revskill10 opening the issue to either document the use-case better or work on an enhancement.

**Comment by frankdugan3** [comment]
I think this is important. One of my use cases would be with an ERP where it is common for users to have multiple roles which have overlapping permissions on the same tables. Permissions in this case are not linear because an admin position for one job area would have less permissions than an operator in another area. I think in that case, the highest permissions matched on user and table/field should be used. For this to work smoothly, I think the results would be expected to be merged as if both roles could be executed at once.

**Comment by coco98** [comment]
@frankdugan3 One of the primary implementation candidates for landing this is to do the merging option as you suggested. 

Please do feel free to DM me in case this feature is a blocker for the project you're working on / evaluating. It would greatly help with our triaging :)
I'm tanmaig on our discord (https://discord.gg/vBPpJkS)

**Comment by PierBover** [comment]
So we've hit this problem too and this is a deal breaker for implementing Hasura on our enterprise system.

The vast majority of role based authorization systems (if not all) allow multiple roles to be assigned to a user. These roles are usually overlapping, so a "customer service" role and a "tech support" role will share some permissions. Not only that, but a user might actually have both roles and in many cases you can't really determine which is the default role since it goes against the idea of having multiple roles.

The solution is very simple, just make the default role optional and use the roles array to check against the configured permissions in Hasura like @frankdugan3 suggested.

Do you have an ETA for this @coco98 ?

**Comment by coco98** [comment]
Hi everyone! 

Would love to hear your thoughts on this related issue: 
[Add role name as root field to graphql schema to make multiple roles easy from the client](#1941)

**Comment by valstu** [comment]
Aren't multiple roles usually "additive" model which means that overlapping permissions will be merged? Lets say you have roles `moderator` (can only delete rows) and 'writer' (can only insert and select rows). If user has both of these roles it can select, delete and insert rows. 

Lets change logic a bit: `moderator` role could select, insert and delete and `writer` role could only select and insert. If user would have role `moderator` and `writer` the writer role would have no effect on permissions because the moderator already have those.

@ecthiender had written the example previously about driver and passenger. I'm not quite sure if I understood the example properly :)

> Now if a user has both the roles passenger and driver, and if the behaviour is to send multiple roles in a query and it applies permissions of all the roles and merges the data and returns, it won't work for the above scenario. In the above scenario, you specifically would want to query as separate roles and not want merged data of both the roles.

If you have both of the roles that means you have access to all data in the table. Then again in reality, `passenger` and `driver` would probably be their own tables with their own permissions and `ride` would be just a cross-reference table. You would probably check if the current user-id matches the `passenger` or `driver` user-id and show the private data on `ride` table based on that. 

But yeah, this isn't easy topic. Do we have investigated how other platforms handle this?

**Comment by PierBover** [comment]
> Lets change logic a bit: moderator role could select, insert and delete and writer role could only select and insert. If user would have role moderator and writer the writer role would have no effect on permissions because the moderator already have those permissions.

I really don't see a problem. Someone wants to delete a row. Authorization logic checks if any of the user roles can do that. Yes? Delete the row. No? Spit the user in the face.

**Comment by evert-smit** [comment]
This merging behaviour would be consistent with 'native' Postgres behaviour for permissive policies:

> When multiple policies apply to a given query, they are combined using either OR (for permissive policies, which are the default) or using AND (for restrictive policies). This is similar to the rule that a given role has the privileges of all roles that they are a member of. Permissive vs. restrictive policies are discussed further below.

https://www.postgresql.org/docs/11/ddl-rowsecurity.html

If a (rough) ETA could be provided, that would be great! Would help to determine if this is nearing implementation.

**Comment by soheilarmin** [comment]
Having merging behavior on roles would help a lot. Then, we can have another header to bypass the merging and then executing the query with a single role or a subset of allowed roles. Or another implementation which is backward compatible.

**Comment by karibertils** [comment]
Running into issues with the current role system and would also like to know if there are any plans/ETA to introduce merged roles ?

**Comment by plwade** [comment]
@coco98 Got any news when this might be included on your sprint? It would be a great enhancement.

**Comment by nolandg** [comment]
Inability to assign multiple roles for a user makes permissions much more work and more error prone. I find myself duplicating a lot of permissions, definitely not DRY design. We need to cascade permissions between roles.

Take for example a blog site. You have content _editors_, _moderators_, and _admins_. Their permissions naturally cascade. A _moderator_ must have all permissions an _editor_ has + more and an admin has all _moderator_ perms + more. The only way I have found to do this with Hasura is to duplicate all the perms from _editor_ over to _moderator_ etc. But then if _editor_ perms change, I have to manually cascade those changes to _moderator_ and _admin_. It becomes a mess. I shouldn't have to say that _admins_ are able to _view_ posts because that has already been accounted for in a "lesser" role.

If we could simply assign _editor, moderator_, and _admin_ to a single user it would be so much more reliable.

**Comment by MarcelloTheArcane** [comment]
Relevant post by Auth0 - https://auth0.com/docs/authorization/concepts/rbac#overlapping-role-assignments

**Comment by cooldesigns** [comment]
Is this issue still not resolved? In a serious application is getting trouble to use hasura, because of this problem.

**Comment by MarcelloTheArcane** [comment]
I have managed to work around this by assigning the `X-Hasura-Role` header to each query/mutation. For example, here is a user and their roles:

![image](https://user-images.githubusercontent.com/21159570/71726377-c95e5f80-2e2e-11ea-8aa1-08bdf7a188e0.png)

Then I can use that role with [`nuxt-community/apollo-module`](https://github.com/nuxt-community/apollo-module)

![image](https://user-images.githubusercontent.com/21159570/71726422-f6127700-2e2e-11ea-8423-60c636fa4994.png)

There are three problems with this approach - 

1. The test if the user is admin (which would not be necessary if [additive authentication](https://github.com/hasura/graphql-engine/issues/877#issuecomment-565343389) was available), and the second is subscriptions.

2. I am using roles with `_read` and `_write` suffixes, and currently I need to duplicate the `_read` permissions over to `_write` in order to access the `returning` field when I mutate something. This isn't so bad, since every user that can write can also read that data.

3. I am using the Apollo module to handle subscriptions, and there doesn't seem to be a way to pass headers on-the-fly for subscriptions. Currently it passes in a Bearer JWT (which has the list of available roles), but nothing else.<br>
  ![image](https://user-images.githubusercontent.com/21159570/71726597-7df88100-2e2f-11ea-8a51-0612fdadeb1c.png)

This is a problem, as it means I have to assign all permissions to the default role (in my case `user`) which compromises my security. I don't want to go down [the same route as @nolandg above](https://github.com/hasura/graphql-engine/issues/877#issuecomment-549997752), and have duplication of permissions.

**Comment by tirumaraiselvan** [comment]
Hey folks, RFC for this is now published here: https://github.com/hasura/graphql-engine/pull/3868

We will take up the implementation as soon as possible.

**Comment by GavinRay97** [comment]
@tirumaraiselvan I am running up against this currently, and I had some ideas about it.

What I did was write a tiny AST visitor that parses GQL queries where roles have been prepended with an identifying symbol (in this case, double underscore `__`) into an array of queries broken down by role.

```js
const { parse, visit, print } = require('graphql/language')

/**
 * Generates an array of objects containing role-query key pairs.
 * 
 * @param {string} query - A graphql query string
 * @returns {Array<{role: string, query: string}>}
 */
function generateRoleSpecificQueries(query) {
  let result = []

  visit(parse(query), {
    enter(node, key, parent) {
      if (node.kind != 'SelectionSet') return
      if (parent.kind != 'Field') return
      if (!parent.name.value.includes('__')) return

      const [_, role] = parent.name.value.split('__')
      const query = `query ${print(node)}`

      const roleSpecificQuery = { role, query }
      result.push(roleSpecificQuery)
    }
  })

  return result
}
```
```js
/**
 *  Checks a query for the presence of roles
 * 
 * @param {string} query - A GraphQL query string
 * @returns {boolean}
 */
function graphqlQueryHasRoles(query) {
  const underscoreRegex = /__\w+/g
  const underscoreWords = query.match(underscoreRegex)
  if (!underscoreWords) return false
  return underscoreWords.some((word) => word != '__typename')
}
```
Now given some query like the following:
```js
const query = `
  query {

    __admin {
      payments {
        id
        amount
      }
    }

    __user {
      messages {
        id
        date
      }
    }
  }`
```

Calling `graphqlQueryHasRoles(query)` will return `true` and calling `generateRoleSpecificQueries(query)` will return
```js
[ { role: 'admin', 
    query: 'query {\n  payments {\n    id\n    amount\n  }\n}' }, 
  { role: 'user', 
    query: 'query {\n  messages {\n    id\n    date\n  }\n}' } ] 
```

So now we have some way to parse role-based queries into individual units. Then the next bit is to implement a method for breaking down each sub-query into requests with the `x-hasura-role` set correctly for each.

The following I have NOT tested, but it should give people a decent idea:

1. Create a middleware/link for Apollo that sets the basis for injecting the `x-hasura-role` header per-request:
```js
const roleLink = new ApolloLink((operation, forward) => {
  operation.setContext({
    headers: {
      'x-hasura-role': operation.getContext().role
    }
  })
  return forward(operation)
})
```
2. Create a wrapper class or something similar over Apollo client, which has a transparent API (by this I mean you still call `client.query()`) that handles the functionality behind parsing requests for roles & sending role-based requests as subqueries:
```js
class HasuraRoleBasedClient {
  constructor(apolloClient) {
    this.client = apolloClient
  }

  async query(queryString) {
    if (graphqlQueryHasRoles(queryString)) {
      let results = []
      const roleQueries = generateRoleSpecificQueries(queryString)
      for (let { role, query } of roleQueries) {
        // Inject the current role into the context for use with
        // getContext() in the role ApolloLink
        const req = await this.client.query({ query, context: { role } })
        // Add the sub-query data to results
        results.push(req)
      }
      // Return results after all role sub-queries completed
      return results
    } else {
      return this.client.query({ query: queryString })
    }
  }
}
```

**Comment by PierBover** [comment]
@tirumaraiselvan is this RFC only for reads? If the RFC is implemented do you have an ETA for writes/deletes/updates?

**Comment by dryhten** [comment]
Hello!

I came across the same issue as well. I basically wanted multiple roles to aggregate permissions since my permissions are simple select/update/insert/delete always with full access to the fields (so I don't have the issue of which role to use when multiple have access to an operation + table).

The way I did it was like this:
1. Created a `getHasuraPermissions` query that will return all hasura permissions for current user along with the roles that grants it. How you implement this is up to what you use (I have a database, but if you don't you'll probably have to get metadata json and get permissions from there - see [here](https://docs.hasura.io/1.0/graphql/manual/migrations/manage-metadata.html)).
2. I store the result on the client as a dictionary of `permissions = { table: { operation: 'role that grants it' } }`. If multiple roles grant the same (operation + table) permission, the first one I come across is picked.
3. On every request I get the role for the operation & table I'm requesting and put that as my `x-hasura-role`. If I have none, it's obviously unauthorized.

This way I don't have to hardcode anything or brute force with each role to see if I get authorized.

Hope this helps somebody!

**Comment by nolandg** [comment]
I'm having trouble extracting from this thread why the RFC is only for reads? Don't all these arguments apply equally for read and write permissions?

**Comment by MatthiasGrandl** [comment]
I have a hard time understanding why not all roles in `x-hasura-allowed-roles` are checked if the default-role doesn't have permission. Wouldn't that solve most issues?

**Comment by mrspartak** [comment]
I hope this will be implemented. I got a really simple example and curious why is this not a case for the most applications.
I got table users and all of them have a public page on my website, so you can check there avatar, nickname etc. Also, there are private data, that only available to the user and currently I ended up with creating role "profile", that I will use only in this place to select more field.
I read this thread and didn't find a reason why there is no check for all the roles in "x-hasura-allowed-roles"

**Comment by michaelwooley** [comment]
Until this is implemented may I request that something along the lines of the [workflow explained](https://github.com/hasura/graphql-engine/issues/877#issuecomment-435269754) by  @go4cas be placed in the docs? (Happy to open as different issue.) After reading the docs front-to-back my understanding of the situation was that any role in `x-hasura-allowed-roles` would still be considered as a possibility when necessary. I ended up spending a ton of time ensuring other parts of my config and rules were correct.

Additionally, for many (but not all) use cases it is possible to get the desired behavior by passing `x-hasura-role` in the request header. Going back and staring at the docs after the fact this is obvious. But not until then.

For example, with apollo client you can do:

```
import { ApolloClient } from 'apollo-client';
import gql from 'graphql-tag';

let client = new ApolloClient(...);
let mutation = gql`...`;
let variables = {...};

let out = await mutate(client, {
  mutation,
  variables,
  context: { headers: { "x-hasura-role": "non-default-role" } }
});
```

**Comment by s-kris** [comment]
Waiting for this to be implemented

**Comment by MrBuddyCasino** [comment]
We're just hitting the same issue. Luckily we can work around this for now, but it feels so dumb. All that is really needed is `x-hasura-allowed-roles`, the rest is kinda... unnecessary?

**Comment by charklewis** [comment]
> I have a hard time understanding why not all roles in `x-hasura-allowed-roles` are checked if the default-role doesn't have permission. Wouldn't that solve most issues?

Is there an update on if this will be implemented?

**Comment by soheilarmin** [comment]
> > I have a hard time understanding why not all roles in `x-hasura-allowed-roles` are checked if the default-role doesn't have permission. Wouldn't that solve most issues?
> 
> Is there an update on if this will be implemented?

IMHO, as roles are not just ON-OFF switches to give/revoke access to objects, this can't be an option. Selecting a role can make a wide range of impact on how/when/what data could be fetched. I'm not sure if it is possible/feasible/logical to automatically select the best role. This should be done on application side. Nested roles is another story which is complicated but logically possible.

**Comment by nolandg** [comment]
I don't see what this _isn't_ a matter of on-off switches. That's exactly what permissions in the console are. I'm not expecting Hasura to "automatically select the best role". Why can't it just perform each operation if _any_ roles are allowed to? This is a basic boolean union of all allowed roles. No one role is "selected". Request arrives, authenticated with a list of allowed roles, if all actions have at least one role that they can be performed under, transaction proceeds, if not, permission denied. 

Is my thinking way off here? I'm quite confused why this thread is so long.

**Comment by s-kris** [comment]
@nolandg you are correct. We should be able to pass roles array for x-hasura-roles and hasura evaluate as union of those.

**Comment by soheilarmin** [comment]
> I don't see what this _isn't_ a matter of on-off switches. That's exactly what permissions in the console are. I'm not expecting Hasura to "automatically select the best role". Why can't it just perform each operation if _any_ roles are allowed to? This is a basic boolean union of all allowed roles. No one role is "selected". Request arrives, authenticated with a list of allowed roles, if all actions have at least one role that they can be performed under, transaction proceeds, if not, permission denied.
> 
> Is my thinking way off here? I'm quite confused why this thread is so long.

Let start with an example that shows us "roles" are not on-off switches.

Role-A: can access "Table-A", with columns 'c1' and 'c2' where 'c1' is greater than 'c2'
BUT
Role-B: can access "Table-A", with columns 'c3' and 'c4' where 'c1' is lower than or equal to 'c2'

now,  tell me if someone has both roles, how this could be done in GraphQL perspective. What types could be selected by the GraphQL query? Let alone joins and relations. This is a very simple example. If we "UNION" these roles as you said, we should be able to return c1,c2,c3,c4 columns for "Table-A". But what about filters? Those rows that are to be selected by "Role-B" could not have columns 'c1' and 'c2'. Should the query return 'null' for those rows while they are not null? 
If you answer these questions, you will understand that roles are NOT on-off switches and it's not logical to mix them.

**Comment by MrBuddyCasino** [comment]
> can access "Table-A", with columns 'c1' and 'c2' where 'c1' is greater than 'c2'

Does anyone actually use this? I mean sure, go make your permission system Turing-complete and stuff starts to get complicated. I have never in 13 years of SW development needed anything more complicated that a standard RBAC, and usually something simpler was fine.

**Comment by soheilarmin** [comment]
> > can access "Table-A", with columns 'c1' and 'c2' where 'c1' is greater than 'c2'
> 
> Does anyone actually use this? I mean sure, go make your permission system Turing-complete and stuff starts to get complicated. I have never in 13 years of SW development needed anything more complicated that a standard RBAC, and usually something simpler was fine.

In my latest 13 years of 24 years of SW development, I'm frequently using much more complicated permissions than this simple example.
Anyway, do you mean that Hasura should eliminate all the amazing stuff of their permission system? I'm not sure this will be an option. That simple example is not actually a use-case. It's just an example to REVEAL that a union on roles is INcorrect by the concept. It does not mean that anyone wants those tables and such funny designs to be used.

**Comment by MrBuddyCasino** [comment]
> Anyway, do you mean that Hasura should eliminate all the amazing stuff of their permission system?

The advanced stuff is nice, the problem is it seems to interfere with the 80% use case that most people have, as the existence of this thread suggests.

**Comment by mitar** [comment]
I think all that can be expressed in SQL, no? So you make all permission queries into SQL, and then combine them with OR. So you have something like `SELECT ... FROM ... WHERE roleACondition OR roleBCondition`. For projections by role, you can also do something like `SELECT CASE roleACondition OR roleBCondition  WHEN true THEN columnC ELSE NULL FROM ....`.

So the question for Hasura is only to properly propagate those role conditions. It is not that you even have to parse that worry about Turing completeness. If your role conditions happen to make it an infinite loop, that is your problem. Hasura just has to pass through conditions.

**Comment by nolandg** [comment]
@soheilarmin What do you mean by "filters"? `where` conditions? 

I guess I don't understand why the permissions system is more cumbersome, painstaking, and error prone for 99% of the real uses cases only to perfectly support some obscure seldom used advanced use cases. As it stands, I can't implement basic RBAC without a lot of maintenance and repetition, very un-DRY. 

If we made the obscure cases the ones that needed some interpretation and the basic cases the ones that just worked as expected, wouldn't it be better? 

I think what we're forgetting is (unless everyone else is coding very different apps from me) that the app and UI always (or at least should) know what data they are allowed to access and mutate. My app knows the current user is only a `moderator` and is not allowed to view credit card information. So when the app requests a row as `moderator` and gets `null` values in the `credit_card_number` column...that's ok. My app knows it doesn't have access to that anyway and my design would never request that in the first place. 

Permissions exist to prevent bad actors from stealing the data, not as a primary means of guiding UI. UI is decoupled from 403 Responses and shouldn't rely on them.

ie. in @soheilarmin's example, yes, return `null` for columns you can't access. No, that's not perfectly consistently correct but the app should know that and shouldn't have even requested them let alone be trusting the response.

If done this way, doesn't it suddenly become easy to implement proper RBAC without the pain?

**Comment by austinrivas** [comment]
As someone who has implemented complex configurable RBAC using hasura that relies on multiple roles this thread reads like FUD.

The hasura team has done a pretty good job of illustrating how to get started down this path here : https://hasura.io/blog/authorization-rules-for-multi-tenant-system-google-cloud/

And how to DRY out your config here: https://hasura.io/blog/using-jsonnet-for-programmable-hasura-metadata/

I would encourage anyone struggling with this concept to invest some time in watching/reading the various conference sessions / blog posts on the topic before coming here and saying what you feel like hasura should / should not be doing.

As amazing as hasura is, it is not a magic bullet. You need to invest some time understanding the product and its scope and evaluating whether that lines up with you goals. Personally I find hasura's system to be extremely flexible and more than enough for my teams needs.

**Comment by revskill10** [comment]
> As someone who has implemented complex configurable RBAC using hasura that relies on multiple roles this thread reads like FUD.
> 
> The hasura team has done a pretty good job of illustrating how to get started down this path here : https://hasura.io/blog/authorization-rules-for-multi-tenant-system-google-cloud/
> 
> And how to DRY out your config here: https://hasura.io/blog/using-jsonnet-for-programmable-hasura-metadata/
> 
> I would encourage anyone struggling with this concept to invest some time in watching/reading the various conference sessions / blog posts on the topic before coming here and saying what you feel like hasura should / should not be doing.
> 
> As amazing as hasura is, it is not a magic bullet. You need to invest some time understanding the product and its scope and evaluating whether that lines up with you goals. Personally I find hasura's system to be extremely flexible and more than enough for my teams needs.

The fact of amazing Hasura features doesn't mean it's enough to cover nessessary real world use cases. In this thread, the focus is on the client side query (multi roles query). Please make the thread focus on the topic instead.

**Comment by austinrivas** [comment]
@revskill10 read the blog posts I linked, it directly discusses how to structure your db for multiple roles.

**Comment by webdeb** [comment]
Waiting for this feature for so long. I've setup a simple rfc some time ago which describes how it could be accomplished with virtual roles: https://github.com/hasura/graphql-engine/pull/5783

A virtual role is a role + check. The idea is you include the virtual role into the query:
```
users_by_pk(id: 123, _as: "me") {
  my_private_field  
}
```

where `me` is a role like **user** + the check `{user_id: {_eq: "x-hasura-user-id"}}`
or for example for posts the virtual role is _author_, you got the idea.

And because virtual roles are just `role + check` you can share them easily, like a generic role `owner` which could be described in your application like **role:user** + **user_id = x-hasura-user-id**

**Comment by lsegal** [comment]
I'm new here but ran into the same issue discussed above and wanted to share some thoughts.

> @revskill10 read the blog posts I linked, it directly discusses how to structure your db for multiple roles.

IMO think the issue is that Hasura is billed as something that layers itself on top of your data in a non-intrusive way and not the other way around. In fact, the headline on Hasura's front page (as of this posting) states "Instant GraphQL [...] for your data". By this you would assume that Hasura takes your data, which includes your schema, and maps it to GraphQL. "Your data", to me, means you get to keep your same data structures, too. Obvious deltas are inevitable, but needing to move columns out of a users table because Hasura's RBAC hadn't thought of the many existing databases that are formulated in this way indicates a pretty big feature gap to me.

The suggestion to "structure data for multiple roles" seems odd, because (a) there are already really good and age-old best practices in schema design that are not compatible with Hasura's feature-set, and (b) it requires you to actively reformulate your own data to fit within Hasura's expectations, often creating heavily normalized structures and can lead to pretty inefficient JOIN queries, especially if you consider the polynomially expanding combinations of data you may need to access across multiple roles. Adding views is easy enough, but this not always possible and also fundamentally assumes that you should just accept breaking changes to your schema or existing data layer / APIs. There seems to be a huge bias towards greenfield data-- what happens when you can't make changes to your API? What happens when you're not in control of your database structure? This is a reality for many existing apps that can't just afford huge changes to existing tables.

I want to double down on the very common case that is [listed above](https://github.com/hasura/graphql-engine/issues/877#issuecomment-716929455) by @webdeb. The "me" user in practice nearly always has different permissions on the users table, including what columns they have access to. To make it even more concrete, such a user might have a `reset_password_token` available, commonly configured on the users table with existing [libraries like Devise](https://github.com/heartcombo/devise/blob/master/lib/devise/models/recoverable.rb#L27-L29). For obvious reasons, that column should only be selectable by `id=X-Hasura-User-Id`, but of course other users should have access to other user fields. There is no differentiation of role applicable here, and Hasura does not support any form of complex JSON configuration that would allow for 2 user roles to access data differently in this way because we're talking about fine grained column grants, not operation access. As such, the only solution is to move data to a private table, but with all the caveats mentioned above, specifically a really large breaking change in both API and data schema, not to mention inheriting complexity onto your backend app, which may or may not even support separating the reset token column from the table (thankfully, [Devise specifically does, but it's not super easy](https://stackoverflow.com/questions/8349509/devise-fields-inside-a-password-table)). There are some other solutions I can think up, but all are extremely awkward and would be alleviated by a full RBAC implementation.

![image](https://user-images.githubusercontent.com/686/99920954-fb451280-2cdb-11eb-9285-b8b243ac22a1.png)

As a sidenote, every single RBAC I've ever used up to Hasura, _including Postgres's own RBAC system_, supports aggregated role grants [via some form of the concept of groups](https://www.postgresql.org/docs/9.0/role-membership.html). This is true in the simplest of systems (UNIX filesystem permissions) and the complex ones (like AWS IAM). The fact that there are concrete use-cases in plain old UNIX filesystem security that requires groups should give us all an indication that it's a very common use case. The idea that needing to have 2 users that can do different things on a given database row (or file on disk) is really not nearly as esoteric as everyone here is trying to make it seem.

TL;DR it would be really great to see some movement on one of the suggestions in this thread. I've only spent a few days with Hasura so far but hit the multiple-role limitation in just a few hours of toying with its RBAC implementation on what I believe is an extremely simple and common use-case. In short, I've never seen an RBAC without some form of hierarchical/grouping support.

**Comment by webdeb** [comment]
@lsegal The link below is even more related to the problem you describe, as it was originally proposed by one of hasuras core members https://github.com/hasura/graphql-engine/issues/3442 Would be cool if we can keep it hot, so it gets some attention, and gets prioritised eventually ;)

**Comment by austinrivas** [comment]
@webdeb your solution of virtual roles basically mirrors what we ended up doing.

We accomplished it by generating our permissions metadata externally to reference a `<resource>_user_roles` table that checks for a given role assignment using the `exists` operator.

**Comment by revskill10** [comment]
I actually develop my own RBAC engine inside Hasura to solve this issue. I'm not sure if this issue is still valid though (my solution doesn't put `roles` in the headers).
By putting `role` in the body inside of `headers`, i could make multiroles query if needed.

**Comment by xiaoyu-tamu** [comment]
@austinrivas

`exists` operator is only for row permission right? How about column permission?

**Comment by webdeb** [comment]
@austinrivas I am very curious, would be cool, if you could share an example, because I am not sure I understood it properly. Also like @xiaoyu-tamu mentioned, are you able to control multiple sets of column permissions for the same role?

**Comment by tirumaraiselvan** [comment]
Released in `v2.0.0-alpha.4` as Inherited Roles. Thanks!

**Comment by mrspartak** [comment]
> Released in `v2.0.0-alpha.4` as Inherited Roles. Thanks!

Any info on how it works?

**Comment by lsegal** [comment]
This is HUGE! Thanks to the team for delivering 🚀

**Comment by tirumaraiselvan** [comment]
Docs here: https://hasura.io/docs/latest/graphql/core/auth/authorization/inherited-roles.html

**Comment by incorvia** [comment]
"Currently, inherited roles are supported only for Postgres read queries and subscriptions." Is there another ticket that tracks this for Mutations and / or Actions?

**Comment by clausMeko** [comment]
Agree with incorvia. The need for mutation inheritance is the most important feature that we need urgently.
